libexec/rtld-elf/aarch64/rtld_machdep.h

```diff
@@ -36,20 +48,121 @@
 #include <sys/types.h>
 #include <machine/atomic.h>
 
+#if __has_feature(capabilities)
+#define MORELLO_FRAG_EXECUTABLE 0x4
+#define MORELLO_FRAG_RWDATA 0x2
+#define MORELLO_FRAG_RODATA 0x1
+#endif
+
 struct Struct_Obj_Entry;
 
 /* Return the address of the .dynamic section in the dynamic linker. */
+#ifdef __CHERI_PURE_CAPABILITY__
+#define rtld_dynamic(obj) (&_DYNAMIC)
+#else
 #define	rtld_dynamic(obj)						\
 ({									\
 	Elf_Addr _dynamic_addr;						\
 	asm volatile("adr	%0, _DYNAMIC" : "=&r"(_dynamic_addr));	\
 	(const Elf_Dyn *)_dynamic_addr;					\
 })
+#endif
 
-Elf_Addr reloc_jmpslot(Elf_Addr *where, Elf_Addr target,
+uintptr_t reloc_jmpslot(uintptr_t *where, uintptr_t target,
     const struct Struct_Obj_Entry *defobj, const struct Struct_Obj_Entry *obj,
     const Elf_Rel *rel);
 
+#if __has_feature(capabilities)
+
+#define	FUNC_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	DATA_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_EXECUTE)
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/* TODO: ABIs with tight bounds */
+#define can_use_tight_pcc_bounds(obj) ((void)(obj), false)
+#endif
+
+/*
+ * Create a pointer to a function.
+ */
+static inline dlfunc_t __capability
+make_code_cap(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj,
+    bool tight_bounds, size_t addend)
+{
+	const void * __capability ret;
+
+	ret = get_codesegment_cap(defobj) + def->st_value;
+	/* Remove store and seal permissions */
+	ret = cheri_clearperm(ret, FUNC_PTR_REMOVE_PERMS);
+	if (tight_bounds) {
+		ret = cheri_setbounds(ret, def->st_size);
+	}
+	/*
+	 * Note: The addend is required for C++ exceptions since capabilities
+	 * for catch blocks point to the middle of a function.
+	 */
+	ret = cheri_incoffset(ret, addend);
+	/* All code pointers should be sentries: */
+	ret = __builtin_cheri_seal_entry(ret);
+	return __DECONST_CAP(dlfunc_t __capability, ret);
+}
+
+/*
+ * Create a function pointer that can be called anywhere
+ */
+static inline dlfunc_t __capability
+make_function_cap_with_addend(const Elf_Sym *def,
+    const struct Struct_Obj_Entry *defobj, size_t addend)
+{
+	/* TODO: ABIs with tight bounds */
+	return make_code_cap(def, defobj, /*tight_bounds=*/false, addend);
+}
+
+static inline dlfunc_t __capability
+make_function_cap(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj)
+{
+	return make_function_cap_with_addend(def, defobj, /*addend=*/0);
+}
+
+static inline void * __capability
+make_data_cap(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj)
+{
+	void * __capability ret;
+	ret = get_datasegment_cap(defobj) + def->st_value;
+	/* Remove execute and seal permissions */
+	ret = cheri_clearperm(ret, DATA_PTR_REMOVE_PERMS);
+	/* TODO: can we always set bounds here or does it break compat? */
+	ret = cheri_setbounds(ret, def->st_size);
```

Comment:
```suggestion
	ret = cheri_setbounds(ret, def->st_size);
```

Reply:
Heh, this was copied from MIPS via RISC-V, will drop the comments from there too as a new commit before this one